File: scripts/enum/sqs.js
'use strict';

const {
  SQSClient,
  ListQueuesCommand,
  GetQueueAttributesCommand,
} = require('@aws-sdk/client-sqs');

const { withRetry, paginate, createLogger } = require('../lib');
const { baseEnum } = require('../lib/base-enum');
const { extractPolicyPrincipals } = require('../lib/policy-parser');

// --- Helpers ---

/**
 * Extracts queue name from URL (last segment after '/').
 */
function queueNameFromUrl(url) {
  if (!url) return null;
  const parts = url.split('/');
  return parts[parts.length - 1];
}

/**
 * Extracts DLQ ARN from RedrivePolicy JSON string.
 */
function extractDlqArn(redrivePolicyJson) {
  if (!redrivePolicyJson) return null;
  try {
    const policy = JSON.parse(redrivePolicyJson);
    return policy.deadLetterTargetArn || null;
  } catch {
    return null;
  }
}

// --- Exported run() for testing ---

async function run(opts = {}) {
  const { runDir, region } = opts;
  const accountId = opts.accountId;

  if (!runDir || !region) {
    throw new Error('runDir and region are required');
  }

  const client = opts.clients?.sqs ?? new SQSClient({ region });

  const logger = opts.logger || createLogger(runDir, 'sqs');
  logger.log('info', 'SQS_Enumeration_Start', { region });

  const findings = [];
  let status = 'complete';
  const errors = [];

  // List all queues (paginated via NextToken)
  let queueUrls;
  try {
    logger.log('api_call', 'ListQueues', { region });
    queueUrls = await paginate(client, ListQueuesCommand, 'QueueUrls', {});
  } catch (err) {
    logger.log('error', 'ListQueues', { error: err.message });
    await logger.flush();
    throw new Error(`ListQueues failed: ${err.message}`);
  }

  // Per-queue: GetQueueAttributes (All)
  for (const queueUrl of queueUrls) {
    const queueName = queueNameFromUrl(queueUrl);

    try {
      logger.log('api_call', 'GetQueueAttributes', { queue: queueUrl });
      const resp = await withRetry(() =>
        client.send(new GetQueueAttributesCommand({
          QueueUrl: queueUrl,
          AttributeNames: ['All'],
        }))
      );
      const attrs = resp.Attributes || {};

      const queueArn = attrs.QueueArn || null;
      const policy = attrs.Policy || null;
      const principals = extractPolicyPrincipals(policy);
      const isFifo = (attrs.FifoQueue === 'true');
      const dlqArn = extractDlqArn(attrs.RedrivePolicy);
      const kmsKeyId = attrs.KmsMasterKeyId || null;
      const visibilityTimeout = parseInt(attrs.VisibilityTimeout, 10) || 30;

      findings.push({
        resource_type: 'sqs_queue',
        resource_id: queueName,
        arn: queueArn,
        region,
        queue_url: queueUrl,
        resource_policy: principals.length > 0 ? { principals } : null,
        fifo: isFifo,
        dlq_arn: dlqArn,
        kms_key_id: kmsKeyId,
        visibility_timeout: visibilityTimeout,
        findings: [],
      });
    } catch (err) {
      errors.push({ resource: queueUrl, error: err.message });
      logger.log('warning', 'GetQueueAttributes', { queue: queueUrl, error: err.message });
    }
  }

  if (errors.length > 0) status = 'partial';

  await logger.flush();
  return { findings, status };
}

if (require.main === module) {
  baseEnum({ module: 'sqs', run });
}

module.exports = { run };

File: scope/enumerators/sqs.py
from __future__ import annotations

import json
from typing import Any

from scope.core.coverage import CoverageTracker
from scope.core.models import ModuleEnvelope
from scope.enumerators.resource_policy import normalize_resource_policy


PRIMARY_CHECKS = ["list_queues"]
REQUIRED_CHECKS = ["queue_attributes"]


def _queue_name_from_url(url: str | None) -> str | None:
    if not url:
        return None
    return url.rstrip("/").split("/")[-1]


def _extract_dlq_arn(redrive_policy: str | None) -> str | None:
    if not redrive_policy:
        return None
    try:
        parsed = json.loads(redrive_policy)
    except json.JSONDecodeError:
        return None
    if isinstance(parsed, dict):
        return parsed.get("deadLetterTargetArn")
    return None


def _extract_policy_principals(policy: str | dict[str, Any] | None) -> list[str]:
    if not policy:
        return []
    try:
        document = json.loads(policy) if isinstance(policy, str) else policy
    except json.JSONDecodeError:
        return []
    statements = document.get("Statement", []) if isinstance(document, dict) else []
    if isinstance(statements, dict):
        statements = [statements]

    principals: set[str] = set()
    for statement in statements:
        if not isinstance(statement, dict) or statement.get("Effect") != "Allow":
            continue
        principal = statement.get("Principal")
        if isinstance(principal, str):
            principals.add(principal)
        elif isinstance(principal, dict):
            for key in ("AWS", "Service", "Federated"):
                value = principal.get(key)
                if isinstance(value, str):
                    principals.add(value)
                elif isinstance(value, list):
                    principals.update(str(item) for item in value)
    return sorted(principals)


def run(factory: Any, region: str) -> ModuleEnvelope:
    tracker = CoverageTracker(primary=PRIMARY_CHECKS, required=REQUIRED_CHECKS)
    sqs = factory.client("sqs")
    findings: list[dict[str, Any]] = []

    try:
        queue_urls = factory.paginate(sqs, "list_queues", "QueueUrls")
        tracker.record_ok("list_queues", resource="queues")
    except KeyError as err:
        if "QueueUrls" in str(err):
            queue_urls = []
            tracker.record_ok("list_queues", resource="queues")
        else:
            tracker.record_module_failure("list_queues", operation="sqs.ListQueues", err=err)
            coverage, errors = tracker.to_envelope_fields()
            return ModuleEnvelope(
                module="sqs",
                account_id=factory.account_id,
                region=region,
                status=tracker.derive_status(),
                resources=[],
                coverage=coverage,
                errors=errors,
            )
    except Exception as err:
        tracker.record_module_failure("list_queues", operation="sqs.ListQueues", err=err)
        coverage, errors = tracker.to_envelope_fields()
        return ModuleEnvelope(
            module="sqs",
            account_id=factory.account_id,
            region=region,
            status=tracker.derive_status(),
            resources=[],
            coverage=coverage,
            errors=errors,
        )

    for queue_url in queue_urls:
        queue_finding = {
            "resource_type": "sqs_queue",
            "resource_id": _queue_name_from_url(queue_url),
            "arn": None,
            "region": region,
            "queue_url": queue_url,
            "resource_policy": None,
            "fifo": False,
            "dlq_arn": None,
            "kms_key_id": None,
            "visibility_timeout": 30,
            "queue_attributes_status": None,
            "findings": [],
        }

        try:
            response = sqs.get_queue_attributes(QueueUrl=queue_url, AttributeNames=["All"])
            attributes = response.get("Attributes", {})
            queue_arn = attributes.get("QueueArn")
            principals = _extract_policy_principals(attributes.get("Policy"))

            queue_finding["arn"] = queue_arn
            queue_finding["resource_policy"] = {"principals": principals} if principals else None
            normalized_policy = normalize_resource_policy(attributes.get("Policy"), account_id=factory.account_id)
            if normalized_policy is not None:
                queue_finding["resource_policy_document"] = normalized_policy["document"]
                queue_finding["resource_policy_statements"] = normalized_policy["statements"]
            queue_finding["fifo"] = attributes.get("FifoQueue") == "true"
            queue_finding["dlq_arn"] = _extract_dlq_arn(attributes.get("RedrivePolicy"))
            queue_finding["kms_key_id"] = attributes.get("KmsMasterKeyId")
            queue_finding["visibility_timeout"] = int(attributes.get("VisibilityTimeout") or 30)
            queue_finding["queue_attributes_status"] = "present"
            tracker.record_ok("queue_attributes", resource=queue_arn or queue_url)
        except Exception as err:
            code = _error_code(err)
            queue_finding["queue_attributes_status"] = (
                "access_denied" if code in {"AccessDenied", "AccessDeniedException"} else "error"
            )
            tracker.record_resource_failure(
                "queue_attributes",
                resource=queue_url,
                operation="sqs.GetQueueAttributes",
                err=err,
            )

        findings.append(queue_finding)

    coverage, errors = tracker.to_envelope_fields()
    return ModuleEnvelope(
        module="sqs",
        account_id=factory.account_id,
        region=region,
        status=tracker.derive_status(),
        resources=findings,
        coverage=coverage,
        errors=errors,
    )


def _error_code(err: Exception) -> str:
    response = getattr(err, "response", None)
    if isinstance(response, dict):
        error = response.get("Error")
        if isinstance(error, dict) and error.get("Code"):
            return str(error["Code"])
    return err.__class__.__name__ or "Error"
